### PR TITLE
AJ-1556: streamline github workflows

### PR DESCRIPTION
**Pull request by davidangb**
Cleans up our GHAs. In this PR:

1. Update `actions/setup-java` to `v4` wherever it is out of date
2. Update `actions/checkout` to `v4` wherever it is out of date
3. Add `gradle/actions/setup-gradle@v3` wherever we use gradle. See https://github.com/marketplace/actions/build-with-gradle for more info; this action was not available when 
4. Remove `setup-java`'s gradle cache in favor of the one in `setup-gradle`
5. Remove explicit caching of gradle stuff via `actions/cache` in favor of the one in `setup-gradle`
6.  Change "Upload Test Reports" steps to only upload reports if tests failed, to save time … does anyone look at these?

### CODE REVIEW

File: .github/workflows/run-e2e-tests.yaml [hunk 1]
```diff
@@ -40,48 +40,48 @@ jobs:
...
         uses: ./.github/actions/landing-zone-attach
         with:
-            run_name: "${{ env.ATTACH_BP_TO_LZ_RUN_NAME }}"
-            bee_name: ${{ env.BEE_NAME }}
-            token: '${{ env.TOKEN }}'
-            project_name: ${{needs.params-gen.outputs.project-name }}
+          run_name: "${{ env.ATTACH_BP_TO_LZ_RUN_NAME }}"
+          bee_name: ${{ env.BEE_NAME }}
+          token: '${{ env.TOKEN }}'
+          project_name: ${{needs.params-gen.outputs.project-name }}
```
Comment (nawatts) [inline .github/workflows/run-e2e-tests.yaml:75]: Super nit, but as long as you're formatting...

```suggestion
          project_name: ${{ needs.params-gen.outputs.project-name }}
```

Same below

File: .github/workflows/build-and-test.yml [hunk 2]
```diff
@@ -80,10 +77,3 @@ jobs:
           SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
         run: ./gradlew --build-cache sonar
-
-      - name: Upload Test Reports
-        if: always()
```
Comment (nawatts) [inline .github/workflows/build-and-test.yml:85]: Might be nice to upload reports on test failure, which would speed up the happy path but still leave them available for debugging.

**Review by nawatts — APPROVED** [review]

File: .github/workflows/run-e2e-tests.yaml [hunk 3]
```diff
@@ -40,48 +40,48 @@ jobs:
...
         continue-on-error: true
         uses: ./.github/actions/landing-zone-attach
         with:
-            run_name: "${{ env.ATTACH_BP_TO_LZ_RUN_NAME }}"
-            bee_name: ${{ env.BEE_NAME }}
-            token: '${{ env.TOKEN }}'
-            project_name: ${{needs.params-gen.outputs.project-name }}
+          run_name: "${{ env.ATTACH_BP_TO_LZ_RUN_NAME }}"
+          bee_name: ${{ env.BEE_NAME }}
+          token: '${{ env.TOKEN }}'
```
Comment (yuliadub) [inline .github/workflows/run-e2e-tests.yaml:74]: stupid question incoming: is there a difference between ' and ", I cant seem to figure it out

Comment (davidangb) [reply .github/workflows/run-e2e-tests.yaml:74]: I doubt there's a difference, I bet this is the result of copy/paste (and I didn't change any of it, just left as-is)

File: .github/workflows/run-e2e-tests.yaml [hunk 4]
```diff
@@ -40,48 +40,48 @@ jobs:
...
 
       - name: dispatch to terra-github-workflows
         id: FirstAttemptCreateBee
         continue-on-error: true
         uses: ./.github/actions/create-bee
         with:
-            bee_name: ${{ env.BEE_NAME }}
-            token: '${{ env.TOKEN }}'
+          bee_name: ${{ env.BEE_NAME }}
+          token: '${{ env.TOKEN }}'
```
Comment (calypsomatic) [inline .github/workflows/run-e2e-tests.yaml:51]: Another weird discrepancy, why does line 51 have quotes but 50 doesn't?

File: .github/workflows/run-e2e-tests.yaml (hunk 3 above)
Comment (nawatts) [reply .github/workflows/run-e2e-tests.yaml:74]: Double quotes allow escape sequences, single quotes only allow printable characters.

https://yaml.org/spec/1.2.2/#single-quoted-style
https://yaml.org/spec/1.2.2/#double-quoted-style

**Review by calypsomatic — APPROVED** [review]
Would mildly prefer standardizing use of ' / " / no quote at all while you're at it, otherwise looks great.

File: .github/workflows/run-e2e-tests.yaml (hunk 1 above)
Comment (davidangb) [reply .github/workflows/run-e2e-tests.yaml]: fixed

File: .github/workflows/build-and-test.yml (hunk 2 above)
Comment (davidangb) [reply .github/workflows/build-and-test.yml:85]: I added these back in, with a condition to only run if tests failed; good call.

**Comment by davidangb** [comment]
> Would mildly prefer standardizing use of ' / " / no quote at all while you're at it, otherwise looks great.

Given that there is actually a behavior difference between these (https://github.com/DataBiosphere/terra-workspace-data-service/pull/513#discussion_r1488440268), I don't really want to take it on in here.

File: .github/workflows/run-e2e-tests.yaml (hunk 3 above)
Comment (yuliadub) [reply .github/workflows/run-e2e-tests.yaml:74]: @nawatts thank you 🙏

**Pull request merged**